command/plugins_install.go

```diff
@@ -128,3 +200,159 @@ func (c *PluginsInstallCommand) RunContext(buildCtx context.Context, args []stri
+
+	var err error
+
+	args.PluginPath, err = filepath.Abs(args.PluginPath)
+	if err != nil {
+		return writeDiags(c.Ui, nil, hcl.Diagnostics{&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Failed to transform path",
+			Detail:   fmt.Sprintf("Failed to transform the given path to an absolute one: %s", err),
+		}})
+	}
+
+	s, err := os.Stat(args.PluginPath)
+	if err != nil {
+		return writeDiags(c.Ui, nil, hcl.Diagnostics{&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Unable to find plugin to promote",
+			Detail:   fmt.Sprintf("The plugin %q failed to be opened because of an error: %s", args.PluginName, err),
+		}})
+	}
+
+	if s.IsDir() {
+		return writeDiags(c.Ui, nil, hcl.Diagnostics{&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Plugin to promote cannot be a directory",
+			Detail:   "The packer plugin promote command can only install binaries, not directories",
+		}})
+	}
+
+	describeCmd, err := exec.Command(args.PluginPath, "describe").Output()
+	if err != nil {
+		return writeDiags(c.Ui, nil, hcl.Diagnostics{&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Failed to describe the plugin",
+			Detail:   fmt.Sprintf("Packer failed to run %s describe: %s", args.PluginPath, err),
+		}})
+	}
+
+	var desc plugin.SetDescription
+	if err := json.Unmarshal(describeCmd, &desc); err != nil {
+		return writeDiags(c.Ui, nil, hcl.Diagnostics{&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Failed to decode plugin describe info",
+			Detail:   fmt.Sprintf("'%s describe' produced information that Packer couldn't decode: %s", args.PluginPath, err),
+		}})
+	}
+
+	semver, err := version.NewSemver(desc.Version)
+	if err != nil {
+		return writeDiags(c.Ui, nil, hcl.Diagnostics{&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid version",
+			Detail:   fmt.Sprintf("Plugin's reported version (%q) is not semver-compatible: %s", desc.Version, err),
+		}})
+	}
+	if semver.Prerelease() != "" {
+		return writeDiags(c.Ui, nil, hcl.Diagnostics{&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid version",
+			Detail:   fmt.Sprintf("Packer can only install plugin releases with this command (ex: 1.0.0), the binary's reported version is %q", desc.Version),
+		}})
+	}
+
+	pluginBinary, err := os.Open(args.PluginPath)
+	if err != nil {
+		return writeDiags(c.Ui, nil, hcl.Diagnostics{&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Failed to open plugin binary",
+			Detail:   fmt.Sprintf("Failed to open plugin binary from %q: %s", args.PluginPath, err),
+		}})
+	}
+
+	pluginContents := &bytes.Buffer{}
+	_, err = io.Copy(pluginContents, pluginBinary)
+	if err != nil {
+		return writeDiags(c.Ui, nil, hcl.Diagnostics{&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Failed to read plugin binary's contents",
+			Detail:   fmt.Sprintf("Failed to read plugin binary from %q: %s", args.PluginPath, err),
+		}})
+	}
+	_ = pluginBinary.Close()
```

Comment:
Maybe we should handle this error in case close fails.

Reply:
Eh, the chances of `Close` failing are excessively thin, and will be closed anyway when the process terminates, so we probably can overlook this one (we would too if we deferred)